File: src/main/scala/adept/utils/Hasher.scala
package adept.utils

import java.io.InputStream
import java.security.MessageDigest

object Hasher {
  private[adept] lazy val md: ThreadLocal[MessageDigest] = new ThreadLocal[MessageDigest] { //make message digest thread-"safe"
    override def initialValue() = {
      MessageDigest.getInstance("SHA-256")
    }
  }

  private def withMd[A](f: MessageDigest => A) = {
    val currentMd = md.get
    currentMd.reset()
    try {
      f(currentMd)
    } finally {
      currentMd.reset()
    }
  }

  def hash(bytes: Array[Byte]): String = withMd { currentMd =>
    currentMd.update(bytes)
    import javax.xml.bind.DatatypeConverter
    DatatypeConverter.printHexBinary(currentMd.digest()).toLowerCase
  }

  def hash(in: InputStream): String = withMd { currentMd =>
    val buf = new Array[Byte](1024 * 4) //_seemed_ to be the fastest when I tried it out when I was writing this
    var len = in.read(buf)

    //streaming was much more efficient than using digest on Array[Byte] - to be verified...
    while (len > 0) {
      currentMd.update(buf, 0, len)
      len = in.read(buf)
    }
    import javax.xml.bind.DatatypeConverter
    DatatypeConverter.printHexBinary(currentMd.digest()).toLowerCase
  }

}

File: adept-lockfile/src/main/java/adept/hash/Hasher.java
package adept.hash;

import java.io.IOException;
import java.io.InputStream;
import java.security.MessageDigest;
import static javax.xml.bind.DatatypeConverter.printHexBinary;

public class Hasher {

  static ThreadLocal<MessageDigest> md = new ThreadLocal<MessageDigest>() {
    @Override
    public MessageDigest initialValue() {
      try {
        return MessageDigest.getInstance("SHA-256");
      } catch (Exception exception) {
        throw new RuntimeException(exception);
      }
    }
  };

  public static String hash(byte[] bytes) {
    MessageDigest currentMd = md.get();
    currentMd.reset();
    try {
      currentMd.update(bytes);
      return printHexBinary(currentMd.digest()).toLowerCase();
    } finally {
      currentMd.reset();
    }
  }

  public static String hash(InputStream is) throws IOException {
    MessageDigest currentMd = md.get();
    currentMd.reset();
    try {
      byte[] buf = new byte[1024 * 4]; // _seemed_ to be the fastest when I
                                       // tried it out when I was writing the
                                       // scala version of this file
      int len = is.read(buf);

      //streaming was much more efficient than using digest on Array[Byte] - to be verified...
      while (len > 0) {
        currentMd.update(buf, 0, len);
        len = is.read(buf);
      }
      return printHexBinary(currentMd.digest()).toLowerCase();
    } finally {
      currentMd.reset();
    }
  }

}
